fondant/pipeline.py

```diff
@@ -0,0 +1,447 @@
+    from kubernetes import client as k8s_client
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class FondantComponentOp(FondantComponentSpec):
+    """
+    Class representing an operation for a Fondant Component in a Kubeflow Pipeline. An operation
+    is a representation of a function that will be executed as part of a pipeline.
+
+    Arguments:
+        component_spec_path: The path to the specification file defining the component
+        arguments: A dictionary containing the argument name and value for the operation.
+        number_of_gpus: The number of gpus to assign to the operation
+        node_pool_name: The name of the node pool to which the operation will be assigned.
+        p_volumes: Collection of persistent volumes in a Kubernetes cluster. Keys are mount paths,
+         values are Kubernetes volumes or inherited types(e.g. PipelineVolumes).
+        ephemeral_storage_request = Used ephemeral-storage request (minimum) for the operation.
+         Defined by string which can be a number or a number followed by one of “E”, “P”, “T”, “G”,
+         “M”, “K”. (e.g. 2T for 2 Terabytes)
+
+    Note:
+        - A Fondant Component operation is created by defining a Fondant Component and its input
+            arguments.
+        - The `number_of_gpus`, `node_pool_name`,`p_volumes` and `ephemeral_storage_size`
+            attributes are optional and can be used to specify additional configurations for
+            the operation. More information on the optional attributes that can be assigned to
+             kfp components here:
+             https://kubeflow-pipelines.readthedocs.io/en/1.8.13/source/kfp.dsl.html
+    """
+
+    component_spec_path: str
+    arguments: t.Dict[str, t.Any]
+    number_of_gpus: t.Optional[int] = None
+    node_pool_name: t.Optional[str] = None
+    p_volumes: t.Optional[t.Dict[str, k8s_client.V1Volume]] = None
+    ephemeral_storage_size: t.Optional[str] = None
+
+    def __post_init__(self):
+        with open(self.component_spec_path, encoding="utf-8") as file_:
+            specification = yaml.safe_load(file_)
+        super().__init__(specification)
+
+    def __name__(self):
+        return self.name
+
+
+class FondantPipeline:
+    """Class representing a Fondant Pipeline."""
+
+    def __init__(
+        self,
+        base_path: str,
+        pipeline_name: str,
+        pipeline_description: t.Optional[str] = None,
+    ):
+        """
+        Initialize the FondantPipeline object.
+
+        Args:
+            base_path: The base path for the pipeline where the artifacts are stored.
+            pipeline_name: The name of the pipeline.
+            pipeline_description: Optional description of the pipeline.
+        """
+        self.base_path = base_path
+        self.name = pipeline_name
+        self.description = pipeline_description
+        self.package_path = f"{pipeline_name}.tgz"
+        self._graph: t.OrderedDict[str, t.Any] = OrderedDict()
+
+    def add_op(
```

Comment:
What happens if more than one op is currently defined without dependencies?

Reply:
I think that's an edge case that I need to address

Reply:
addressed and added some tests